Add generated schema registry client

nguerrera opened the pull request:
Initializes `@azure/service-registry` with scaffolding from template/ folder and generated autorest  code.

No outer convenience layer is implemented yet, nothing is exported from the library, and no artifacts are set to publish. We will implement the outer layer in separate PR(s) in the September milestone.

Fixes #10076

nguerrera commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,124 @@
+{
+  "name": "@azure/schema-registry",
+  "version": "1.0.8",
```
What version should go here?

nguerrera commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,124 @@
+{
+  "name": "@azure/schema-registry",
+  "version": "0.1.0",
```
I wasn't sure what version to put here.

xirzec replied:
```suggestion
  "version": "1.0.0-preview.1",
```

this is traditional, though we are thinking of changing from `preview` to `prerelease`

ramya-rao-a replied:
We can go with `1.0.0-preview.1` for now. This package will take another month to ship, we can change later to prerelease if needed

nguerrera replied:
Done.

xirzec commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,124 @@
+  "main": "dist/index.js",
+  "module": "dist-esm/src/index.js",
+  "browser": {
+    "stream": "./node_modules/stream-browserify/index.js",
```
probably don't need either of these mappings?

xirzec commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,124 @@
+    "@azure/core-tracing": "1.0.0-preview.8",
+    "@azure/logger": "^1.0.0",
+    "@opentelemetry/api": "^0.6.1",
+    "events": "^3.0.0",
```
can ditch this dependency too

xirzec commented on sdk/schemaregistry/schema-registry/review/schema-registry.api.md:
```diff
@@ -0,0 +1,96 @@
+
+// @public (undocumented)
+export class SchemaRegistryClient extends SchemaRegistryClientContext {
+    constructor($host: string, options?: SchemaRegistryClientOptionalParams);
```
`host` param having a `$` prefix is weird to me

ramya-rao-a replied:
@nguerrera,  Is the $ coming from the swagger? I don't recall our code generator deciding to add this on its own.

nguerrera replied:
It was, but it's now named endpoint (no $) as of latest swagger.

xirzec commented on sdk/schemaregistry/schema-registry/review/schema-registry.api.md:
```diff
@@ -0,0 +1,96 @@
+}
+
+// @public
+export interface SchemaRegistryClientRegisterSchemaHeaders {
```
this interface shows up a lot under different names - should be just one interface I think?

ramya-rao-a replied:
Good feedback to send back to the swagger PR

nguerrera replied:
cc @arerlend

arerlend replied:
@jhendrixMSFT is there a way to create definitions for header groups in swagger 2?  I believe it is in v3.

nguerrera replied:
(If it's not possible with v2, it won't be blocking as we can ensure no duplication like this in what we actually export.)

jhendrixMSFT replied:
Sadly no, this is something not in v2.

xirzec commented on sdk/schemaregistry/schema-registry/review/schema-registry.api.md:
```diff
@@ -0,0 +1,96 @@
+
+// @public
+export type SchemaRegistryClientGetIdBySchemaContentResponse = SchemaRegistryClientGetIdBySchemaContentHeaders & SchemaId & {
+    _response: coreHttp.HttpResponse & {
```
is it useful to return the `_response` property? should we elide it?

nguerrera replied:
(Artifact of generation, stopped exporting generated code)

xirzec commented on sdk/schemaregistry/schema-registry/src/index.ts:
```diff
@@ -0,0 +1,6 @@
+// Licensed under the MIT license.
+
+export * from "./generated/models";
+export { SchemaRegistryClient } from "./generated/schemaRegistryClient";
```
the generated client should be wrapped in a convenience layer with telemetry, reference doc comments, etc.

nguerrera replied:
Stopped exporting generated code

xirzec commented on sdk/schemaregistry/schema-registry/review/schema-registry.api.md:
```diff
@@ -0,0 +1,96 @@
+// @public (undocumented)
+export class SchemaRegistryClient extends SchemaRegistryClientContext {
+    constructor($host: string, options?: SchemaRegistryClientOptionalParams);
+    getIdBySchemaContent(groupName: string, schemaName: string, serializationType: string, schemaContent: string, options?: coreHttp.OperationOptions): Promise<SchemaRegistryClientGetIdBySchemaContentResponse>;
```
since both this API and `registerSchema` have the same four required string properties, I think I'd rather have this be an object shape:

```suggestion
    getIdBySchemaContent(schema: SchemaDescription, options?: coreHttp.OperationOptions): Promise<SchemaRegistryClientGetIdBySchemaContentResponse>;
```

and then elsewhere

```ts
interface SchemaDescription {
  groupName: string;
  schemaName: string;
  serializationType: string;
  schemaContent: string;
}

nguerrera replied:
I forwarded this feedback to the apiview for the Java client that is further along. We can match the outcome of that review when we build the outer layer over the generated code.

xirzec commented on sdk/schemaregistry/schema-registry/review/schema-registry.api.md:
```diff
@@ -0,0 +1,96 @@
+    constructor($host: string, options?: SchemaRegistryClientOptionalParams);
+    getIdBySchemaContent(groupName: string, schemaName: string, serializationType: string, schemaContent: string, options?: coreHttp.OperationOptions): Promise<SchemaRegistryClientGetIdBySchemaContentResponse>;
+    getSchemaById(schemaId: string, options?: coreHttp.OperationOptions): Promise<SchemaRegistryClientGetSchemaByIdResponse>;
+    registerSchema(groupName: string, schemaName: string, serializationType: string, schemaContent: string, options?: coreHttp.OperationOptions): Promise<SchemaRegistryClientRegisterSchemaResponse>;
```
is `serializationType` the payload format e.g. `json`/`avro`?

nguerrera replied:
I believe so. cc @arerlend

arerlend replied:
Yup there is a bit of flakiness on whether we are referring to this encoding format, e.g. JSON or Avro, as 'serialization type' or 'schema type'.  Would be good to make a choice on Monday at the arch-board review so we can make a final call.

nguerrera commented:
@xirzec Thanks! This is great feedback towards the final shape of the API. 

So far, as you saw, this is entirely a result of the auto-generation from the current swagger. My direction with this PR was to follow how I interpreted #10076 defining a checkpoint for an autogenerated package that we wouldn't ship yet. (I might have misinterpreted.)

> This work is planned for the August milestone and does not involve shipping the package which will be done in the September milestone.
>
>decide package name
>auto generate the package using the latest v6 code generator on the swagger
>set up CI

I was imagining anything involving wrapping the generation into a convenient API that conforms to our guidelines to come in follow up PR(s) for the September milestone.

I could also build the better layer from the get go and hold this PR back until I have it. I don't see much downside to that. It would likely mean missing the August 7th date, but I'm not sure that's meaningful if we weren't going to ship what's in on that date anyway.

Happy to adjust as the team sees fit. @ramya-rao-a thoughts?

nguerrera commented:
Another approach would maybe be to not even export the generated code in this first PR, just to make it clearer that this layer is not actually intended to ever be directly consumable (beyond also not shipping it :)).

(Opportunistically, this would help with an immediate problem where api-extractor has started disliking how I'm doing that, as of the latest swagger, which I'm currently trying to debug.)

xirzec commented:
Ah, thank you for explaining. I'm fine with either proposed solution: not exporting the client in this PR or having a follow-up to provide the actual convenience shape.

xirzec commented:
One minor note: if you put a ci.yml in your namespace folder, it means you're signing up to publish nightlies of this package, so if the name isn't final don't add ci.yml yet.

ramya-rao-a commented:
> I was imagining anything involving wrapping the generation into a convenient API that conforms to our guidelines to come in follow up PR(s) for the September milestone.

That is right

> I could also build the better layer from the get go and hold this PR back until I have it. 

Not required. We can build the convenience layer in a separate PR.

> Another approach would maybe be to not even export the generated code in this first PR, just to make it clearer that this layer is not actually intended to ever be directly consumable (beyond also not shipping it :)).

I like this.

> One minor note: if you put a ci.yml in your namespace folder, it means you're signing up to publish nightlies of this package, so if the name isn't final don't add ci.yml yet.

Is there no way to set up nightly test runs, but without the publishing the nightly package? The idea here was to keep all the infra ready so that when we do add tests, that is all we would need to focus on.

xirzec commented:
> Is there no way to set up nightly test runs, but without the publishing the nightly package? The idea here was to keep all the infra ready so that when we do add tests, that is all we would need to focus on.

Hmm I think as long as we remove the artifact portion it'll do what you want, see https://github.com/Azure/azure-sdk-for-js/pull/9277

nguerrera commented:
OK. I've removed the exporting of the generated code and removed the artifacts from ci.yml as discussed above, addressed other feedback, and taken this out of draft.

ramya-rao-a commented on sdk/schemaregistry/schema-registry/rollup.base.config.js:
```diff
@@ -0,0 +1,103 @@
+const production = process.env.NODE_ENV === "production";
+
+export function nodeConfig(test = false) {
+  const externalNodeBuiltins = ["events"];
```
We don't depend on `events`, so we can keep this as an empty array as a placeholder for future external node builtins

```suggestion
  const externalNodeBuiltins = [];
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/rollup.base.config.js:
```diff
@@ -0,0 +1,103 @@
+      }),
+      cjs({
+        namedExports: {
+          events: ["EventEmitter"],
```
We don't depend on `events`, so this can be removed too

```suggestion
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,121 @@
+    "@azure/core-http": "^1.1.4",
+    "@azure/core-tracing": "1.0.0-preview.8",
+    "@azure/logger": "^1.0.0",
+    "@opentelemetry/api": "^0.6.1",
```
We currently don't use any of the above dependencies other than `@azure/core-http`.
But, these will come into play when we start adding the convenience layer and tracing features, so we should be good keeping them

nguerrera replied:
I think I'll remove what's not used now. Easy to add back referencing template, and better to avoid any risk of shipping with unused dependencies.

ramya-rao-a commented on sdk/schemaregistry/schema-registry/package.json:
```diff
@@ -0,0 +1,121 @@
+    "rollup-plugin-terser": "^5.1.1",
+    "rollup-plugin-visualizer": "^4.0.4",
+    "typescript": "~3.9.3",
+    "util": "^0.12.1"
```
There are way too many dev dependencies here, some of which are not used even in the template project. Logged #10391 for follow up

nguerrera replied:
Tried to trim it down.

ramya-rao-a commented on sdk/schemaregistry/schema-registry/sample.env:
```diff
@@ -0,0 +1,17 @@
+# Used in most samples. Retrieve these values from a Cognitive Services instance
+# in the Azure Portal.
+ENDPOINT="Your Endpoint URL"
+API_KEY="Your API key"
```
We usually would like to preface service specific env vars with the service name to differentiate from similar env vars of other services.
And API_KEY is not applicable to SchemaRegistry.

```suggestion
SCHEMA_REGISTRY_ENDPOINT="Your Endpoint URL"
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/karma.conf.js:
```diff
@@ -0,0 +1,129 @@
+
+    envPreprocessor: [
+      "TEST_MODE",
+      "ENDPOINT",
+      "API_KEY",
+      "API_KEY_ALT",
```
We usually would like to preface service specific env vars with the service name to differentiate from similar env vars of other services.
And API_KEY is not applicable to SchemaRegistry.

```suggestion
      "SCHEMA_REGISTRY_ENDPOINT",
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/javascript/package.json:
```diff
@@ -0,0 +1,30 @@
+  "homepage": "https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/schema-registry/schema-registry",
+  "sideEffects": false,
+  "dependencies": {
+    "@azure/identity": "latest",
```
Should also add the package the sample is meant to show case :)


```suggestion
    "@azure/identity": "latest",
    "@azure/schema-registry":  "latest"
```

nguerrera replied:
```suggestion
    "@azure/identity": "latest",
    "@azure/schema-registry":  "latest",
```
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/javascript/sample.env:
```diff
@@ -0,0 +1,13 @@
+# Used in most samples. Retrieve these values from a Cognitive Services instance
+# in the Azure Portal.
+ENDPOINT="<Your Endpoint URL>"
+API_KEY="<Your API key>"
```
```suggestion
SCHEMA_REGISTRY_ENDPOINT="<Your Endpoint URL>"
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/javascript/schemaRegistrySample.js:
```diff
@@ -0,0 +1,27 @@
+const dotenv = require("dotenv");
+dotenv.config();
+
+// You will need to set this environment variables or edit the following values
+// const endpoint = process.env["ENDPOINT"] || "<endpoint>";
+const apiKey = process.env["API_KEY"] || "<api key>";
```
```suggestion
// You will need to set this environment variables or edit the following values
const endpoint = process.env["SCHEMA_REGISTRY_ENDPOINT"] || "<endpoint>";
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/typescript/package.json:
```diff
@@ -0,0 +1,39 @@
+  "homepage": "https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/schema-registry/schema-registry",
+  "sideEffects": false,
+  "dependencies": {
+    "@azure/identity": "latest",
```
```suggestion
    "@azure/identity": "latest",
    "@azure/schema-registry": "latest",
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/typescript/sample.env:
```diff
@@ -0,0 +1,13 @@
+# Used in most samples. Retrieve these values from a Cognitive Services instance
+# in the Azure Portal.
+ENDPOINT="Your Endpoint URL"
+API_KEY="<Your API key>"
```
```suggestion
SCHEMA_REGISTRY_ENDPOINT="Your Endpoint URL"
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/samples/typescript/src/schemaRegistrySample.ts:
```diff
@@ -0,0 +1,27 @@
+dotenv.config();
+
+// You will need to set this environment variables or edit the following values
+// const endpoint = process.env["ENDPOINT"] || "<endpoint>";
+const apiKey = process.env["API_KEY"] || "<api key>";
```
```suggestion
const endpoint = process.env["SCHEMA_REGISTRY_ENDPOINT"] || "<endpoint>";
```

ramya-rao-a commented on sdk/schemaregistry/schema-registry/src/generated/schemaRegistryClientContext.ts:
```diff
@@ -0,0 +1,56 @@
+import { SchemaRegistryClientOptionalParams } from "./models";
+
+const packageName = "@azure/schema-registry";
+const packageVersion = "1.0.0";
```
This needs to match the package version in package.json file.

```suggestion
const packageVersion = "1.0.0-preview.1";
```

ramya-rao-a commented on sdk/schemaregistry/ci.yml:
```diff
@@ -0,0 +1,29 @@
+      - sdk/schemaregistry/
+
+pr:
+  branches:
```
@mitchdenny The branches info here and in the above section are common to all Track 2 packages. Any reason why this needs to be repeated for all packages and cannot be in the parent yml file?

nguerrera commented on sdk/schemaregistry/ci.yml:
```diff
@@ -0,0 +1,29 @@
+# DO NOT EDIT THIS FILE
+# This file is generated automatically and any changes will be lost.
```
For my education, when/how is this generated? There has been manual edits to this since copied from template to get this library bootstrapped. Is there a better way to do that?

ramya-rao-a replied:
I am not sure about this one.
@KarishmaGhiya ?

xirzec replied:
The comment is a lie currently. We hand author these AFAIK.

nguerrera replied:
Should we have an issue to remove these comments, then?

nguerrera replied:
I filed #10434
